app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.test.js

```diff
@@ -0,0 +1,521 @@
+import { ethErrors } from 'eth-rpc-errors';
+import {
+  CHAIN_IDS,
+  NETWORK_TYPES,
+} from '../../../../../shared/constants/network';
+import addEthereumChain from './add-ethereum-chain';
+
+const NON_INFURA_CHAIN_ID = '0x123456789';
+
+const MOCK_MAINNET_CONFIGURATION = {
+  chainId: CHAIN_IDS.MAINNET,
+  nickname: 'Ethereum Mainnet',
+  rpcUrl: 'https://mainnet.infura.io/v3/',
+  type: NETWORK_TYPES.MAINNET,
+  ticker: 'ETH',
+  rpcPrefs: {
+    blockExplorerUrl: 'https://etherscan.io',
+  },
+};
+
+const MOCK_OPTIMISM_CONFIGURATION = {
+  chainId: CHAIN_IDS.OPTIMISM,
+  nickname: 'Optimism',
+  rpcUrl: 'https://optimism.llamarpc.com',
+  rpcPrefs: {
+    blockExplorerUrl: 'https://optimistic.etherscan.io',
+  },
+  ticker: 'ETH',
+};
+
+const MOCK_NON_INFURA_CONFIGURATION = {
+  chainId: NON_INFURA_CHAIN_ID,
+  rpcUrl: 'https://custom.network',
+  ticker: 'CUST',
+  nickname: 'Custom Network',
+  rpcPrefs: {
+    blockExplorerUrl: 'https://custom.blockexplorer',
+  },
+};
+
+describe('addEthereumChainHandler', () => {
+  const addEthereumChainHandler = addEthereumChain.implementation;
+
+  const setupMocks = ({
+    permissionedChainIds = [],
+    permissionsFeatureFlagIsActive,
+    overrides = {},
+  } = {}) => {
+    return {
+      getChainPermissionsFeatureFlag: () => permissionsFeatureFlagIsActive,
+      getCurrentChainIdForDomain: jest
+        .fn()
+        .mockReturnValue(NON_INFURA_CHAIN_ID),
+      setNetworkClientIdForDomain: jest.fn(),
+      findNetworkConfigurationBy: jest
+        .fn()
+        .mockReturnValue(MOCK_MAINNET_CONFIGURATION),
+      setActiveNetwork: jest.fn(),
+      getCurrentRpcUrl: jest
+        .fn()
+        .mockReturnValue(MOCK_MAINNET_CONFIGURATION.rpcUrl),
+      requestUserApproval: jest.fn().mockResolvedValue(123),
+      requestSwitchNetworkPermission: jest.fn(),
+      getCaveat: jest.fn().mockReturnValue({ value: permissionedChainIds }),
+      upsertNetworkConfiguration: jest.fn().mockResolvedValue(123),
+      startApprovalFlow: () => ({ id: 'approvalFlowId' }),
+      endApprovalFlow: jest.fn(),
+      ...overrides,
+    };
+  };
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('when getChainPermissionsFeatureFlag() returns false', () => {
+    describe('if a networkConfiguration for the given chainId does not already exist', () => {
+      it('should call requestApproval, mockUpsertNetworkConfiguration with the requested chain, and setActiveNetwork', async () => {
```

Comment:
There are a number of `describe` blocks in this file that wrap a single test case. Instead, we could try to distill the test case description:
```suggestion
    it('creates a new network configuration for the given chain id if none exists', async () => {
```

There a number of cases like this in the file, and you we should consider addressing all of them.

Reply:
Regarding the nested describes yeah it got a bit out of hand haha.

Regarding the different naming style you propose here I wonder if the imperative approach I'm taking, highlighting what hooks should be called is more clear than the declarative approach or whether its unnecessary since the testing assertions themselves achieve this more effectively

Reply:
reduced the nesting here and made the test names more declarative: https://github.com/MetaMask/metamask-extension/pull/24415/commits/8eb1076ad57063a5029b3add5ff05f518826da3e